tensorboard/program.py

```diff
@@ -647,9 +647,15 @@ def init(wsgi_app, flags):
         max_attempts = 100 if should_scan else 1
         base_port = min(base_port + max_attempts, 0x10000) - max_attempts
 
+        def is_port_in_use(port):
+            with socket.socket(socket.AF_INET, socket.SOCK_STREAM) as s:
+                return s.connect_ex(("localhost", port)) == 0
+
         for port in range(base_port, base_port + max_attempts):
             subflags = argparse.Namespace(**vars(flags))
             subflags.port = port
+            if is_port_in_use(port):
```

Comment:
Doing the port-in-use check here won't respect the `should_scan` logic (based on `flags.port`).

Can we instead move it closer to where the logic previously happened (when binding to the socket would raise the exception)? I.e. move it into `WerkzeugServer.__init__()` before the call to the superclass init (which is where we'll attempt binding the socket) and make it raise TensorBoardPortInUseError, which will then trigger the existing `except` logic on the lines below.

Reply:
This portion of the code is a bit complicated so while I think I've made the suggested change, please let me know if I've missed something.